tests/python/pants_test/backend/codegen/tasks/test_simple_codegen_task.py

```diff
@@ -0,0 +1,230 @@
+    SimpleCodegenTaskTest.DummyGen.ALL_TARGETS = target_roots
+    return self._create_subclass_task(SimpleCodegenTaskTest.DummyGen,
+                                      target_roots=target_roots,
+                                      **options)
+
+  def test_codegen_strategy(self):
+    task = self.create_task(self.context())
+
+    self.set_options_for_scope(task.options_scope, strategy='global')
+    task = self.create_task(self.context())
+    self.assertEqual('global', task.get_options().strategy)
+    self.assertEqual('global', task.codegen_strategy.name)
+
+    self.set_options_for_scope(task.options_scope, strategy='isolated')
+    task = self.create_task(self.context())
+    self.assertEqual('isolated', task.codegen_strategy.name)
+
+    class TestGen(SimpleCodegenTask):
+      """Minimal class to test that forced codegen strategy works as intended."""
+
+      @property
+      def forced_codegen_strategy(self):
+        return 'global'
+
+    task = self._create_subclass_task(TestGen, strategy='global')
+    self.assertEqual('global', task.codegen_strategy.name)
+    task = self._create_subclass_task(TestGen, strategy='isolated')
+    self.assertEqual('global', task.codegen_strategy.name)
+
+  def test_codegen_workdir_suffix(self):
+    target_specs = [
+      'project/src/main/foogen/foo-lib:foo-target-a',
+      'project/src/main/foogen/foo-lib:foo-target-b',
+      'project/src/main/foogen/foo-bar:foo-target-a',
+      'project/src/main/genfoo/foo-bar:foo-target-a',
+    ]
+    spec_paths = set([spec.split(':')[0] for spec in target_specs])
+    spec_names = set([spec.split(':')[1] for spec in target_specs])
+    for spec_path in spec_paths:
+      parent_dir = os.path.dirname(spec_path)
+      if not os.path.exists(parent_dir):
+        self.create_dir(parent_dir)
+      self.create_file('{0}/BUILD'.format(spec_path), '\n'.join([dedent('''
+          dummy_library(name='{name}',
+            sources=[],
+          )
+        '''.format(name=spec_name)) for spec_name in spec_names]))
+
+    targets = set([self.target(spec) for spec in target_specs])
+
+    def get_suffix(target, strategy):
+      return SimpleCodegenTask._codegen_strategy_for_name(strategy)(None)\
+          .codegen_workdir_suffix(target)
+
+    for target in targets:
+      self.assertEqual('global', get_suffix(target, 'global'))
+      self.assertTrue('isolated' in get_suffix(target, 'isolated'))
+
+    global_dirs = set(get_suffix(target, 'global') for target in targets)
+    isolated_dirs = set(get_suffix(target, 'isolated') for target in targets)
+
+    self.assertEqual(1, len(global_dirs), 'There should only be one global directory suffix!')
+    self.assertEqual(len(targets), len(isolated_dirs),
+                     'There should be exactly one directory suffix per unique target!')
+
+    for target in targets:
+      for strategy in ('global', 'isolated',):
+        suffix = None
+        for i in range(5):
```

Comment:
Yeah, you're right, the test is a bit too paranoid. I was trying to make it robust in case a subclass did something really subtle and stupid, but it's probably more sensible to keep the test simple and try to catch sneaky instability in code review.